tests/src/topology/mod.rs

```diff
@@ -255,4 +261,287 @@ impl Topology {
     pub fn executors(&self) -> &[Executor] {
         &self.executors
     }
+
+    pub async fn wait_network_ready(&self) {
+        const POLL_INTERVAL: Duration = Duration::from_millis(200);
+        const TIMEOUT: Duration = Duration::from_secs(60);
+
+        let listen_ports = self.node_listen_ports();
+        if listen_ports.len() <= 1 {
+            return;
+        }
+
+        let initial_peer_ports = self.node_initial_peer_ports();
+        let expected_peer_counts = find_expected_peer_counts(&listen_ports, &initial_peer_ports);
+        let labels = self.node_labels();
+
+        let readiness = timeout(adjust_timeout(TIMEOUT), async {
+            loop {
+                let validator_infos =
+                    join_all(self.validators.iter().map(Validator::network_info)).await;
+
+                let executor_infos =
+                    join_all(self.executors.iter().map(Executor::network_info)).await;
+
+                let infos: Vec<_> = validator_infos
+                    .into_iter()
+                    .chain(executor_infos.into_iter())
+                    .collect();
+
+                let all_ready = infos
+                    .iter()
+                    .enumerate()
+                    .all(|(idx, info)| info.n_peers >= expected_peer_counts[idx]);
+
+                if all_ready {
+                    break;
+                }
+
+                sleep(POLL_INTERVAL).await;
+            }
+        });
+
+        if readiness.await.is_err() {
+            let validator_infos =
+                join_all(self.validators.iter().map(Validator::network_info)).await;
+
+            let executor_infos = join_all(self.executors.iter().map(Executor::network_info)).await;
+
+            let infos: Vec<_> = validator_infos
+                .into_iter()
+                .chain(executor_infos.into_iter())
+                .collect();
+
+            let summary = build_timeout_summary(&labels, infos, &expected_peer_counts);
+
+            panic!("timed out waiting for network readiness: {summary}");
+        }
+    }
+
+    pub async fn wait_membership_ready(&self) {
+        self.wait_membership_assignations_non_empty(SessionNumber::from(0u64))
+            .await;
+    }
+
+    pub async fn wait_membership_assignations_non_empty(&self, session: SessionNumber) {
+        self.wait_membership_ready_with(session, "non-empty assignations", |resp| {
+            !resp.assignations.is_empty()
+        })
+        .await;
+    }
+
+    pub async fn wait_membership_assignations_empty(&self, session: SessionNumber) {
+        self.wait_membership_ready_with(session, "empty assignations", |resp| {
+            resp.assignations.is_empty()
+        })
+        .await;
+    }
+
+    pub async fn wait_membership_ready_with<F>(
```

Comment:
Can't `wait_network_ready` be specialized version of this since they are (or look so) the same code?

Reply:
I think it can be done, but if its too complicated no need.